source/common/ssl/context_config_impl.cc

```diff
@@ -105,6 +105,11 @@ unsigned ContextConfigImpl::tlsVersionFromProto(
 ClientContextConfigImpl::ClientContextConfigImpl(
     const envoy::api::v2::auth::UpstreamTlsContext& config)
     : ContextConfigImpl(config.common_tls_context()), server_name_indication_(config.sni()) {
+  // We need to check the length of the C string in case there are embedded \000
+  // that BoringSSL will treat as end-of-string. PGV can't do this.
+  if (!server_name_indication_.empty() && strlen(server_name_indication_.c_str()) == 0) {
```

Comment:
That will only work if `\0` is the first byte.

If you want to go C route, then you most likely want:

```
if (!server_name_indication_.empty() &&
    ::memchr(server_name_indication_.c_str(), 0, server_name_indication_.length()) != nullptr) {
  throw EnvoyException("SNI names containing NULL-byte are not allowed");
}
```

Otherwise, I think that `std::string::find('\0')` should also work, i.e.:

```
if (server_name_indication_.find('\0') != std::string::npos) {
  throw EnvoyException("SNI names containing NULL-byte are not allowed");
}
```

Reply:
Why do we care about the presence of the NULL byte? We care about two things:
1. Is the string effectively empty if this is interpreted as a C string.
2. Is the string too long.

Reply:
Because if there is a NULL-byte in proto (e.g. `sni: exam\0ple.com`), then we're effectively telling BoringSSL to set SNI to a value different (e.g. `exam`) than the one specified in the configuration.

Also, the check for "does it contain NULL-byte?" is pretty much the same as "does it start with NULL-byte?", so I don't see a reason not to do that.

Reply:
I agree with @PiotrSikora. We should reject any string with a \0 in it, because BoringSSL won't use anything after the \0.